src/pool/mempool.rs

```diff
@@ -21,95 +33,146 @@ pub type KakarotPool<Client> = Pool<Validator<Client>, TransactionOrdering, Noop
 
 #[derive(Debug, Default)]
 pub struct AccountManager {
-    accounts: HashSet<Felt>,
+    accounts: Arc<Mutex<HashMap<Felt, Felt>>>,
 }
 
 impl AccountManager {
-    pub fn new(path: &str) -> Self {
-        let mut accounts = HashSet::new();
+    pub async fn new<SP>(path: &str, eth_provider: &EthDataProvider<SP>) -> eyre::Result<Self>
+    where
+        SP: starknet::providers::Provider + Send + Sync + 'static,
+    {
+        let mut accounts = HashMap::new();
 
         // Open the file specified by `path`
         let Ok(mut file) = File::open(path) else {
-            return Self::default();
+            return Ok(Self::default());
         };
 
         let mut contents = String::new();
         if file.read_to_string(&mut contents).is_err() {
-            return Self::default();
+            return Ok(Self::default());
         }
 
         // Parse the file contents as JSON
         let json: Value = match serde_json::from_str(&contents) {
             Ok(json) => json,
             Err(_) => {
-                return Self::default();
+                return Ok(Self::default());
             }
         };
 
         // Extract the account addresses from the JSON array
         if let Some(array) = json.as_array() {
             for item in array {
-                if let Some(address) = item.as_str() {
-                    accounts.insert(Felt::from_hex_unchecked(address));
+                if let Some(account_address) = item.as_str() {
+                    let felt_address = Felt::from_hex_unchecked(account_address);
+
+                    let starknet_block_id = eth_provider
+                        .to_starknet_block_id(Some(BlockId::default()))
+                        .await
+                        .map_err(|e| eyre::eyre!("Error converting block ID: {:?}", e))?;
+
+                    // Query the initial account_nonce for the account from the provider
+                    accounts.insert(
+                        felt_address,
+                        eth_provider
+                            .starknet_provider()
+                            .get_nonce(starknet_block_id, felt_address)
+                            .await
+                            .unwrap_or_default(),
+                    );
                 }
             }
         }
 
```

Comment:
we should probably error out if `accounts.len() == 0`

Reply:
It's about the nonce here. I think we use this approach elsewhere in the code to say that if there is an error then we set the default nonce. In other parts of the code we check if the contract is not found or if the entrypoint is not found but I think here it is overkill right? We can just say that we return the default nonce (0) knowing that it should never happen.

Reply:
oh I don't mean about the nonce, I mean at the end of the loop, we should probably error out if we didn't find any accounts in the json file

Reply:
Oh sure just added it